relay-server/src/services/processor/playstation.rs

```diff
@@ -2,42 +2,328 @@
+        for log_line in prospero_dump.log_lines {
+            console_log.push_str(log_line);
+        }
+        if !console_log.is_empty() {
+            let mut item = Item::new(ItemType::Attachment);
+            item.set_filename("console.log");
+            item.set_payload(ContentType::Text, console_log.into_bytes());
+            item.set_attachment_type(AttachmentType::Attachment);
+            envelope.add_item(item);
+        }
 
         metric!(counter(RelayCounters::PlaystationProcessing) += 1);
     }
 
-    Ok(())
+    Ok(Some(EventFullyNormalized(false)))
 }
 
-pub fn process(
-    _managed_envelope: &mut TypedEnvelope<ErrorGroup>,
-    _event: &mut Annotated<Event>,
-) -> Result<Option<EventFullyNormalized>, ProcessingError> {
-    // TODO: Add the processing logic here.
-    Ok(None)
+pub fn update_sentry_event(event: &mut Event, prospero: &ProsperoDump) {
+    let contexts = event.contexts.get_or_insert_with(Contexts::default);
+
+    let mut device_context = DeviceContext {
+        name: prospero
+            .system_name
+            .map(|s| Annotated::new(s.to_string()))
+            .unwrap_or_default(),
+        arch: Annotated::new("x86_64".into()),
+        model_id: prospero
+            .hardware_id
+            .clone()
+            .map(Annotated::new)
+            .unwrap_or_default(),
+        other: [("manufacturer".into(), Annotated::new("Sony".into()))].into(),
+        ..Default::default()
+    };
+
+    let mut os_context = OsContext {
+        name: Annotated::new("PlayStation".into()),
+        ..Default::default()
+    };
+    let mut runtime_context = RuntimeContext::default();
+
+    if let Some(hardware_id) = prospero.hardware_id.clone() {
+        event.server_name = Annotated::new(hardware_id);
+    }
+
+    // Elevate userdata values with sentry. prefix into the sentry event.
+    // Set the release based on the "release" key in the userdata
+    if let Some(release) = prospero
+        .userdata
+        .get("sentry.release")
+        .or_else(|| prospero.userdata.get("release"))
+    {
+        event.release = Annotated::new(LenientString(release.clone().into_owned()));
+    }
+    if let Some(environment) = prospero.userdata.get("sentry.environment") {
+        event.environment = Annotated::new(environment.clone().into_owned());
+    }
+    if let Some(username) = prospero.userdata.get("sentry.user.username") {
+        event.user.get_or_insert_with(Default::default).username =
+            Annotated::new(LenientString(username.to_string()));
+    }
+    if let Some(email) = prospero.userdata.get("sentry.user.email") {
+        event.user.get_or_insert_with(Default::default).email = Annotated::new(email.to_string());
+    }
+
+    let tags = event.tags.value_mut().get_or_insert_with(Tags::default);
+    macro_rules! add_tag {
+        ($key:expr, $value:expr) => {
+            tags.push(Annotated::new(TagEntry(
+                Annotated::new($key),
+                Annotated::new($value),
```

Comment:
If you move the `.into()`/`.to_owned()` calls into the macro you'll have to type less.

Here `.into()` makes sense, because you are in 'generic'(-like) context and can now use the macro with different types and they will convert correctly.